### DISCUSSION TITLE
How to generate a specified amount of `State`s and `StatesGroup`s

### DISCUSSION THREAD

**Discussion by Salz0** [aiogram/aiogram#657] [Q&A] [answered]
Hey guys, I want to be able to create something like a metaclass for states

Purpose:
I want to be able to create an infinite amount of commands/responses, so I need to have a generalized solution to creating and applying states to a DIALOG.

Perfectly -> run a for loop for a predefined range (f.e. 4) and get the number of unique states that I need (in this case I'd have 4 unique states created)

Ready to buy a coffee for the one who solves this ;)

**Comment by evgfilim1** [comment]
If I understood you correctly, you can create your states as usual and loop them inside handlers like this:

```python
class States(StatesGroup):
    a = State()
    b = State()
    c = State()

@dp.message_handler(commands=['start'])
@dp.message_handler(state=States.a)
async def first(msg):
    await msg.reply('Gone to second')
    await States.b.set()


@dp.message_handler(state=States.b)
async def first(msg):
    await msg.reply('Gone to third')
    await States.c.set()


@dp.message_handler(state=States.c)
async def first(msg):
    await msg.reply('Gone to start')
    await States.a.set()
```

**Reply by Salz0** [reply]
Almost,

Point is to create something like this:

```py
class MetaState(StatesGroup):
    ...

for i in range(x):
    @dp.message_handler(commands=[f'start{i}'])
    @dp.message_handler(state=MetaState.lambda_state)  # Question is how to loop those things in a similar fashion. How do I generate and x amount of states, or different states in one loop?
    async def x_state(msg):
        await msg.reply(f'Gone to state {x}')
        await MetaState.next()
```

I need to be able to create any amount of states as I wish to create any amount of dialogs

**Reply by evgfilim1** [reply]
`StateFilter` supports accepting not only one `State`, but also `Collection[State]` and even full `StatesGroup` as an argument.
https://github.com/aiogram/aiogram/blob/580fa2e49913e68412eae51840fa73aeb2e9fa36/aiogram/dispatcher/filters/builtin.py#L508-L532

**Reply by Salz0** [reply]
Didn't quite get how I can loop through a collection and assign a state to a list of handlers. 
Could you specify?

**Reply by Salz0** [reply]
Maybe a bit clearer -> how do I create a collection of states? Without hardcoding it, as a general function

```py

def any_collection(amount) -> f"Collection with {amount} States":
    ...

**Comment by Salz0** [comment, marked as answer]
Ok, so I came up with the answer myself. If anyone will search for an answer:

```py
def generate_state_groups(
    group_prefix: str, groups_amount: int, states_amount: int
) -> list:
    """
    Generates any amount of unique groups with a strict amount of states for each.
    """
    return [
        type(
            f"{group_prefix}_{n}",
            (StatesGroup,),
            {f"{n}state{nn}": State() for nn in range(states_amount)},
        )
        for n in range(groups_amount)
    ]
```

**Reply by shakhrillodev** [reply]
Where is my coffee? :)
